### PR TITLE
Ryan/feat/multi stage/14 skip diffusion

### PR DESCRIPTION
**Pull request by ryanontheinside**
Adds `skip_diffusion` mode to run pipeline pre/post-processors without the diffusion step, taking advantage of the hook/module system.  
Also improves the CUDA stream implementation from #115 and makes the RealESRGAN processor more robust in this context.  

> Note: Models are still loaded when using `skip_diffusion`. A true “minimal pipeline” option is out of scope for this PR and can be added later once model loading/management is more robust.

### CODE REVIEW

File: src/streamdiffusion/preprocessing/postprocessing_orchestrator.py [hunk 1]
```diff
@@ -184,8 +205,10 @@ def _apply_single_postprocessor(self,
             processed_tensor = input_tensor.to(device=self.device, dtype=self.dtype)
             
             # CRITICAL: Convert from VAE output range [-1,1] to processor input range [0,1]
+            print(f"_apply_single_postprocessor: Input tensor range: min={processed_tensor.min().item():.4f}, max={processed_tensor.max().item():.4f}")
```
Comment (varshith15) [inline src/streamdiffusion/preprocessing/postprocessing_orchestrator.py:208]: nit: is this print needed?

File: src/streamdiffusion/preprocessing/postprocessing_orchestrator.py [hunk 2]
```diff
@@ -184,8 +205,10 @@ def _apply_single_postprocessor(self,
...
             logger.debug(f"_apply_single_postprocessor: Converting tensor from VAE range [-1,1] to processor range [0,1]")
             processor_input = (processed_tensor / 2.0 + 0.5).clamp(0, 1)
+            print(f"_apply_single_postprocessor: Processor input range: min={processor_input.min().item():.4f}, max={processor_input.max().item():.4f}")
```
Comment (varshith15) [inline src/streamdiffusion/preprocessing/postprocessing_orchestrator.py:211]: nit: is this print needed?

File: src/streamdiffusion/preprocessing/postprocessing_orchestrator.py [hunk 3]
```diff
@@ -203,9 +226,22 @@ def _apply_single_postprocessor(self,
             
             # Ensure result is a tensor
             if isinstance(processor_output, torch.Tensor):
+                # Check for NaN corruption
+                if processor_output.isnan().any():
+                    print(f"_apply_single_postprocessor: NaN CORRUPTION DETECTED in processor output!")
+                    logger.error(f"_apply_single_postprocessor: Processor output contains NaN values - memory corruption detected")
```
Comment (varshith15) [inline src/streamdiffusion/preprocessing/postprocessing_orchestrator.py:232]: this isnt doing anything?

File: src/streamdiffusion/preprocessing/processors/realesrgan_trt.py [hunk 4]
```diff
@@ -161,23 +176,30 @@ def __init__(self, enable_tensorrt: bool = True, force_rebuild: bool = False, **
...
-            
-            self._engine = RealESRGANEngine(str(self.engine_path))
-            self._engine.load()
-            self._engine.activate()
-            
-            # Allocate buffers for standard input size (will be reallocated as needed)
-            standard_shape = (1, 3, 512, 512)
-            self._engine.allocate_buffers(standard_shape, device=self.device)
+            with self._engine_lock:
+                # Double-check locking pattern
```
Comment (varshith15) [inline src/streamdiffusion/preprocessing/processors/realesrgan_trt.py:191]: this logic isnt written right, why is there an `self._engine is None` and then a lock and then `self._engine is None` again?

File: src/streamdiffusion/wrapper.py [hunk 5]
```diff
@@ -513,13 +521,15 @@ def update_stream_params(
         ipadapter_config : Optional[Dict[str, Any]]
             IPAdapter configuration dict containing scale, style_image, etc.
         """
+        # skip_diffusion = True
```
Comment (varshith15) [inline src/streamdiffusion/wrapper.py:524]: nit: cleanup

File: src/streamdiffusion/pipeline.py [hunk 6]
```diff
@@ -106,6 +106,9 @@ def __init__(
         # Default IP-Adapter runtime weight mode (None = uniform). Can be set to strings like
         # "ease in", "ease out", "ease in-out", "reverse in-out", "style transfer precise", "composition precise".
         self.ipadapter_weight_type = None
+        
+        # Skip diffusion runtime flag - when True, bypasses VAE encode/decode and diffusion
+        self.skip_diffusion = False
```
Comment (varshith15) [inline src/streamdiffusion/pipeline.py:111]: why is this needed? why add it to pipeline?

File: src/streamdiffusion/wrapper.py [hunk 7]
```diff
@@ -555,11 +565,76 @@ def __call__(
         Union[Image.Image, List[Image.Image]]
             The generated image.
         """
+        if getattr(self.stream, 'skip_diffusion', False):
```
Comment (varshith15) [inline src/streamdiffusion/wrapper.py:568]: just have skip_diffusion as part of wrapper not pipeline and stream_parameter_updater

File: src/streamdiffusion/wrapper.py [hunk 8]
```diff
@@ -555,11 +565,76 @@ def __call__(
...
+        Union[Image.Image, List[Image.Image], torch.Tensor, np.ndarray]
+            The processed image with hooks applied.
+        """
+
+        #TODO: add safety checker call somewhere in this method
+
+
+        if self.mode == "txt2img":
+            logger.warning("_process_skip_diffusion: skip_diffusion mode not applicable for txt2img - no input image")
+            return self.txt2img(prompt)
```
Comment (varshith15) [inline src/streamdiffusion/wrapper.py:605]: nit: raising an error would be better?

File: src/streamdiffusion/wrapper.py [hunk 9]
```diff
@@ -555,11 +565,76 @@ def __call__(
...
+        else:
+            processed_tensor = image
+        
+        # Apply image preprocessing hooks (expect [0,1] range - pre-VAE encoding)
+        # Convert [-1,1] -> [0,1] for preprocessing hooks
+        preprocessor_input = self._denormalize_on_gpu(processed_tensor)
+        preprocessor_output = self.stream._apply_image_preprocessing_hooks(preprocessor_input)
+        
+        # Convert [0,1] -> [-1,1] back to pipeline range for postprocessing hooks
+        processed_tensor = self._denormalize_from_processors(preprocessor_output)
```
Comment (varshith15) [inline src/streamdiffusion/wrapper.py:630]: idk if all these are needed, just check and remove the redundant sutff

File: src/streamdiffusion/wrapper.py [hunk 10]
```diff
@@ -873,6 +954,19 @@ def _load_model(
...
+        
+        # Reset CUDA context to prevent corruption from previous runs
+        try:
+            if torch.cuda.is_available():
+                torch.cuda.empty_cache()
+                torch.cuda.synchronize()
+                # Force CUDA context reset by creating and destroying a small tensor
+                temp_tensor = torch.zeros(1, device=self.device)
+                del temp_tensor
+                torch.cuda.empty_cache()
```
Comment (varshith15) [inline src/streamdiffusion/wrapper.py:966]: dont need to emoty_cache multiple times, just once at the end

File: src/streamdiffusion/preprocessing/processors/realesrgan_trt.py (hunk 4 above)
Comment (ryanontheinside) [reply src/streamdiffusion/preprocessing/processors/realesrgan_trt.py:191]: You're absolutely right, the double-checked locking pattern was unnecessarily complex for this use case. I've simplified it to just a single check without the lock.

**Comment by BuffMcBigHuge** [comment]
LGDM. 🚀
